### ISSUE TITLE
Nested sections in dashboard

### ISSUE THREAD

**Issue by bearinsun** [vapid/vapid#12]
Is it possible to have nested sections in the dashboard?

For example, say there is an `index.html` file which specifies a root section with the name "index." Currently, this would generate a page in the dashboard called "Index," with all child sections also receiving their own page in the dashboard. Could Vapid instead generate children/sub sections inside of the root "index" page of the dashboard?
 
```html
{{#section index}
    <h1>{{title}}</h1>
    <ul>
        {{#section items}}
        <li>{{name}}</li>
        {{/section}}
    </ul>
{{/section}}
```

**Comment by trst** [comment]
I tried the same thing earlier and ran into the same error. I do think this would be helpful for organising the sections of Vapid's backend but is there any other particular reasons for nesting beyond that?

I am asking honestly, it's been a long day 😆

**Comment by bearinsun** [comment]
The main use case for this, although I'm sure there are more that I'm not thinking of, would be for multiple pages.

If a site has "About," "Projects," and "Home" pages, it would be hugely beneficial if each HTML page could represent a tab in the dashboard, with each section on that page acting as a group for related fields inside of that tab.

With this, there's another level of categorization that, while less useful to the developer(s), is especially useful to the users in charge of updating content. The ability to have the fields organized by page, and then by sections is extremely useful in that regard. Not to mention that without it, it's difficult to define similar sections (e.g. partials) across multiple pages, as they don't appear (correct me if I am wrong) to take the scope of their parent section in the dashboard.

Vapid is an awesome project, a great idea, and its almost the perfect use case for a project that @Nexuist and I have been working on, but unfortunately nested sections in the manner described above are a necessary feature before we can consider moving forward with using it.

**Comment by bearinsun** [comment]
@srobbin would you be open to a PR for this feature?

@Nexuist and I are considering putting some time into getting nested sections working as it's something we're looking to utilize relatively soon. Do you have a rough idea of what work it would entail?

**Comment by srobbin** [comment]
@bearinsun: Right now, I'd prefer to keep the sections flat, and not introduce nesting. Although I'm grateful that you're interested in contributing.

My two main reasons being:

1. It would make it more confusing to differentiate between nested sections and broader reusable content. For example, if you wanted to include office locations on the About and Contact page:

```
<!-- about.html -->
{{#section about}}
  <h1>About</h1>
  {{#section people}}
     <!-- a subsection, profile images and bios -->
  {{/section}}

  {{#section offices}}
     <!-- content that exists on multiple pages -->
  {{/section}}
{{/section}}

<!-- contact.html -->
{{#section contact}}
  <h1>Contact</h1>
  {{#section offices}}
     <!-- content that exists on multiple pages -->
  {{/section}}
{{/section}}
```

2. I'd like to give Vapid some time and see how people use it first. i.e., Are flat sections an impediment, or will people find creative, acceptable ways to embrace the constraint?

**Comment by bearinsun** [comment]
Thank you for taking the time to leave a detailed response, @srobbin.

While I understand the reasoning behind keeping sections flat for the time being, it admittedly does make applying Vapid for our use case somewhat more complicated. Our users (the people who edit the site's content and who would be using Vapid's dashboard) are accustomed to an organizational hierarchy based on individual pages, further categorized by their sections. 

For example, on the site we would like to use Vapid for, each page contains a banner image, title and caption. Our current dashboard reflects exactly this, providing a tab for each page and then a "header" section inside each tab where the image, title, and caption can be entered.

What is the "Vapid approach" to this sort of situation? Is there an alternative solution that better fits Vapid's feature set as it stands today and that does not compromise categorical capabilities?

**Comment by srobbin** [comment]
@bearinsun I think the Vapid approach could be something like this: each section representing a page, and associated items appearing adjacent to the parent page in the nav (since sections are ordered alphabetically). For example:

```
# about.html
{{#section about_page}}
  {{header_image type=image}}
  {{header_title}}
  {{header_caption type=html}}
  {{other}}
  {{fields}}
{{/section}}
```

```
# features.html
{{#section features_page}}
  {{header_image type=image}}
  {{header_title}}
  {{header_caption type=html}}
  {{something_else}}a
{{/section}}
```

```
# services.html
{{#section services_page}}
  {{header_image type=image}}
  {{header_title}}
  {{header_caption type=html}}
  
  {{#section services_items}}
    {{title}}
    {{body type=html}}
  {{/section}}

  {{another_thing}}
{{/section}}
```

Becomes:

![screenshot](http://g.recordit.co/aeA9qy8zSj.gif)

**Comment by bearinsun** [comment]
@srobbin I see... thanks for demonstrating. This approach works excellently if every page has just one or two sections, but it begins to break down as more sections are added.

If the "Services" page has a list of purchasable services, contacts, and, say, gallery of images, that would entail four separate tabs for each section on the dashboard. Take this example and add even just two or three more pages with their own sections and the number of dashboard tabs would rapidly get out of hand, leaving a lot of repeated text (in the tab names) and making it difficult to sort through.

I would wager that this is a situation that most sites will eventually run into, at least for anything more complex than a small blog or portfolio site. If these sort of sites are Vapid's target demographic, _that's perfectly fine_ and in that case the decision to hold off on implementing nested sections is understandable.

If not, I want to kindly ask that you please reconsider this as a core feature. For an enhancement that's relatively small (although I can't speak on the complexity of the change), it offers a lot in return for both the developer and end user. Given that Vapid is intended "for people who build websites for other people," I think that it would make Vapid all the more applicable and powerful for a wider range of projects people could use it for.

<details>
	<summary>
		<sup>Semi-related sidenote</sup>
	</summary>
	<sub>
		It seems Vapid uses 
		<a href="https://semantic-ui.com/" target="_noblank" rel="noopener noreferrer">Semantic UI</a>, which offers a 
		<a href="https://semantic-ui.com/collections/menu.html" target="_noblank" rel="noopener noreferrer">Menu component</a> and 
		<a href="https://semantic-ui.com/modules/tab.html" target="_noblank" rel="noopener noreferrer">Tab component</a>, both of which could work well for navigating between nested
		sections in a dashboard page...
	</sub>
</details>

**Comment by Nexuist** [comment]
@srobbin here is a visual representation of how nested sections work in the (rough and outdated) editor @bearinsun and I originally built (for the site we hope to transition to Vapid). If it is not distinguishable (we know its design is less than perfect—it was a quick and dirty implementation at the time), "Qualities" is the nested section in the screenshot.

Note that for just the "Home" page, there are four different sections (not all visible in the image) where multiple items are required. The other five pages of the site also have several sections. Vapid's approach applied to this content would create a haphazard dashboard for the content editors.

![Editor image](https://i.imgur.com/2Igubzo.png)

**Comment by srobbin** [comment]
@Nexuist, @bearinsun: Thank you—this is good feedback. I'll add improving section management to the roadmap.

**Comment by moqmar** [comment]
I've got a syntax idea for that:

```html
{{#section city multiple=true}}
  <h1>{{name}}</h1>
  <ul>
    {{#section city.office multiple=true}}
      <li>
        <h2>{{name}}</h2>
        <address>{{street}}, {{^name}}</address>
      </li>
    {{/section}}
  </ul>
```

Let me explain:
- The section `city.office` contains all offices. If it's used inside a `city` section (nested in the template), it only contains the offices inside that city.
- We print the name of the parent section (the city) using `^name`. If we have more deeply nested sections, we can use e.g. `^^^name`, to use the parent's parent's parent's name.
- If we want to display all offices regardless of the city, but inside the city section, we can use `^city.office` as a section.
- If we have more than two section layers (e.g. `city.office.department`), and are inside `city.office`, we can use `^city.office.department` for all departments inside that city (instead of the office), and `^^city.office.department` for all departments.

We can now do stuff like the following:
```
{{#section city.office.department multiple=true}}
  {{#unless ^closed}}<li>{{name}}</li>{{/unless}}
{{/section}}
```

This prints every department that's not in a closed office. We're using nested sections, but are not nesting them inside the template.

The dashboard for that could look like this:
![grafik](https://user-images.githubusercontent.com/5559994/47956082-79bca380-df9f-11e8-98a9-173ccb796e06.png)

Basically, showing all of the current nesting level (and maybe above), and what's inside. If I click "Office" now, I can see all offices in Berlin. If I click "New York", I can see that it has a subsection "Office".

If that suggestion seems to make sense, I can try to create a pull request for that?

**Comment by megalofauna** [comment]
I understand the desire to keep sections flat and see what comes of it, but I immediately ran into the desire to nest sections on my very first Vapid site. Specifically, I wanted to have a repeating {{#section}} inside a non-repeating {{#section}}. 

Loving Vapid. But miss the equivalent functionality from Perch.

**Comment by Voydfox** [comment]
Just picked up Vapid, looking to implement it into a couple projects I'm working on. I too was hoping for the ability to nest sections, has any thought been put towards it? Thanks! <3
